File: src/main/java/ru/simpleGBD/App/Logic/Proxy/AbstractProxyListProvider.java
package ru.simpleGBD.App.Logic.Proxy;

import com.google.api.client.http.HttpResponse;
import ru.simpleGBD.App.Config.GBDOptions;
import ru.simpleGBD.App.Logic.ExecutionContext;
import ru.simpleGBD.App.Utils.HttpConnections;
import ru.simpleGBD.App.Utils.Logger;

import java.net.InetSocketAddress;
import java.util.*;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

/**
 * Created by km on 27.11.2015.
 */
@SuppressWarnings("ALL")
public abstract class AbstractProxyListProvider implements IProxyListProvider {

    private static final String DEFAULT_PROXY_DELIMITER = ":";

    private static final Logger logger = Logger.getLogger(ExecutionContext.output, "ProxyPistProvider");

    private static IProxyListProvider INSTANCE;

    final Set<HttpHostExt> proxyList = new HashSet<>();
    private boolean proxyListCompleted = false;
    List<String> proxyItems;

    private HttpHostExt processProxyItem(String proxyItem) {
        HttpHostExt proxy = null;

        try {
            String[] proxyItemArr = splitItems(proxyItem);

            if (proxyItemArr == null || proxyItemArr.length < 2)
                return null;

            InetSocketAddress host = new InetSocketAddress(proxyItemArr[0], Integer.parseInt(proxyItemArr[1]));
            String cookie = getCookie(host);
            if (cookie != null) {
                proxy = new HttpHostExt(host, cookie);

                if ((GBDOptions.secureMode() && proxy.isSecure()) || !GBDOptions.secureMode()) {
                    proxyList.add(proxy);
                    logger.info(String.format("%sroxy %s added.",
                            GBDOptions.secureMode() ? proxy.isSecure() ? "Secure p" : "NOT secure p" : "P",
                            host.toString()));
                } else
                    logger.info(String.format("NOT secure proxy %s NOT added.", host.toString()));
            } else
                logger.info(String.format("Proxy %s NOT added.", host.toString()));
        } catch (Exception ex) {
            logger.info(String.format("Not valid proxy string %s.", proxyItem));
        }

        return proxy;
    }

    private String[] splitItems(String proxyItem, String delimiter) {
        return proxyItem.split(delimiter);
    }

    private String[] splitItems(String proxyItem) {
        String[] tmpItems = splitItems(proxyItem, DEFAULT_PROXY_DELIMITER);
        if (tmpItems != null && tmpItems.length >= 2)
            return tmpItems;
        else {
            tmpItems = splitItems(proxyItem, "\\s+");
            return tmpItems != null && tmpItems.length >= 2 ? tmpItems : null;
        }
    }

    private String getCookie(InetSocketAddress proxy) {
        try {
            HttpResponse resp = HttpConnections.getResponse(proxy);

            return ((ArrayList) resp.getHeaders().get("set-cookie")).get(0).toString().split(";")[0];
        } catch (Exception e) {
            return null;
        }
    }

    class InternalProxyIterator implements Iterator<HttpHostExt> {

        private final Iterator<String> itr;

        private InternalProxyIterator() {
            itr = proxyItems.iterator();

            proxyListCompleted = true;
        }

        @Override
        public boolean hasNext() {
            return itr.hasNext();
        }

        @Override
        public HttpHostExt next() {
            return processProxyItem(itr.next());
        }
    }

    @Override
    public Iterator<HttpHostExt> getProxyList() {
        return proxyListCompleted ? proxyList.iterator() : new InternalProxyIterator();
    }

    public static IProxyListProvider getInstance() {
        if (INSTANCE == null)
            INSTANCE = GBDOptions.getProxyListFile() == null ? new StaticProxyListProvider() : new WebProxyListProvider();

        return INSTANCE;
    }

    @Override
    public void invalidatedProxyListener() {
        long liveProxyCount = proxyList.stream().filter(p -> p.isAvailable()).count();
        if (liveProxyCount == 0 && GBDOptions.secureMode())
            throw new RuntimeException("No more proxies!");
    }

    @Override
    public Stream<HttpHostExt> getParallelProxyStream() {
        Iterator<HttpHostExt> hostIterator = getProxyList();
        Iterable<HttpHostExt> iterable = () -> hostIterator;
        return StreamSupport.stream(iterable.spliterator(), true);
    }
}

File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/proxy/providers/AbstractProxyListProvider.kt
package ru.kmorozov.gbd.core.logic.proxy.providers

import com.google.common.base.Strings
import ru.kmorozov.gbd.core.config.GBDOptions
import ru.kmorozov.gbd.core.logic.context.ExecutionContext
import ru.kmorozov.gbd.core.logic.proxy.HttpHostExt
import ru.kmorozov.gbd.core.logic.proxy.UrlType
import ru.kmorozov.gbd.core.logic.proxy.web.WebProxyListProvider
import ru.kmorozov.gbd.logger.Logger
import ru.kmorozov.gbd.utils.HttpConnections
import ru.kmorozov.gbd.utils.QueuedThreadPoolExecutor
import ru.kmorozov.gbd.utils.SetBlockingQueue
import java.net.InetSocketAddress
import java.util.*
import java.util.concurrent.LinkedBlockingQueue
import java.util.stream.Stream

/**
 * Created by km on 27.11.2015.
 */

abstract class AbstractProxyListProvider : IProxyListProvider {

    override val proxyList: LinkedBlockingQueue<HttpHostExt> = SetBlockingQueue()
    protected var proxyItems: MutableSet<Optional<InetSocketAddress>> = HashSet()

    init {
        initList()
    }

    open fun initList() {}

    override val parallelProxyStream: Stream<HttpHostExt>
        get() = proxyList.parallelStream()

    override val proxyCount: Int
        get() = proxyList.size

    fun getSomeProxy(): HttpHostExt {
        if (proxyList.isEmpty()) {
            val proxy = proxyList.take()
            proxyList.add(proxy)
            return proxy
        } else
            return proxyList.element()
    }

    private fun splitItems(proxyItem: String): Array<String> {
        var tmpItems = splitItems(proxyItem, DEFAULT_PROXY_DELIMITER)
        if (2 <= tmpItems.size)
            return tmpItems
        else {
            tmpItems = splitItems(proxyItem, "\\s+")
            return if (2 <= tmpItems.size) tmpItems else emptyArray()
        }
    }

    override fun processProxyList(urlType: UrlType) {
        ExecutionContext.proxyExecutor = QueuedThreadPoolExecutor(proxyItems.size, 5, { true }, "proxyExecutor")
        proxyItems.forEach {
            ExecutionContext.proxyExecutor.execute {
                val opHost = processProxyItem(it.get(), urlType)
                if (opHost.isPresent)
                    ExecutionContext.sendProxyEvent(opHost.get())
            }
        }

        ExecutionContext.sendProxyEvent(HttpHostExt.NO_PROXY)
    }

    override fun invalidatedProxyListener() {
        proxyList.removeIf { !it.isAvailable }
        val liveProxyCount = proxyList.count()
        if (0 == liveProxyCount) {
            logger.severe("No more proxies!")
            ExecutionContext.INSTANCE.forceCompleteAll()
        }
    }

    protected fun getInetAddress(proxyItem: String): Optional<InetSocketAddress> {
        val proxyItemArr = splitItems(proxyItem)
        if (2 > proxyItemArr.size)
            return Optional.empty()
        else
            return Optional.of(InetSocketAddress(proxyItemArr[0], Integer.parseInt(proxyItemArr[1])))
    }

    private fun processProxyItem(host: InetSocketAddress, urlType: UrlType): Optional<HttpHostExt> {
        val proxy: HttpHostExt

        val cookie = HttpConnections.INSTANCE.getCookieString(host, urlType)
        proxy = HttpHostExt(host, cookie)

        if (!Strings.isNullOrEmpty(cookie)) {
            if (!GBDOptions.secureMode || proxy.isSecure) {
                logger.finest("${if (GBDOptions.secureMode) if (proxy.isSecure) "Secure p" else "NOT secure p" else "P"}roxy $host added.")
            } else {
                logger.finest("NOT secure proxy $host NOT added.")
                proxy.forceInvalidate(false)
            }
        } else {
            logger.finest("Proxy $host NOT added.")
            proxy.forceInvalidate(false)
        }

        if (proxy.isAvailable) {
            proxyList.add(proxy)
            return Optional.of(proxy)
        } else
            return Optional.empty()
    }

    protected fun notBlacklisted(proxyStr: String): Boolean {
        return !ProxyBlacklistHolder.BLACKLIST.isProxyInBlacklist(proxyStr)
    }

    fun reset() {
        proxyList.removeIf { !it.isAvailable }
    }

    abstract fun findCandidates()

    companion object {

        private const val DEFAULT_PROXY_DELIMITER = ":"

        private val logger = Logger.getLogger(GBDOptions.debugEnabled, AbstractProxyListProvider::class.java)

        val INSTANCE: AbstractProxyListProvider
            get() = if (Strings.isNullOrEmpty(GBDOptions.proxyListFile))
                EmptyProxyListProvider.INSTANCE
            else if (GBDOptions.proxyListFile.equals("web", ignoreCase = true))
                WebProxyListProvider.INSTANCE
            else if (GBDOptions.proxyListFile.equals("tor", ignoreCase = true))
                TorProxyListProvider.INSTANCE
            else FileProxyListProvider.INSTANCE

        fun updateBlacklist() {
            if (INSTANCE.proxyList.size > 1)
                ProxyBlacklistHolder.BLACKLIST.updateBlacklist(INSTANCE.proxyList)
        }

        private fun splitItems(proxyItem: String, delimiter: String): Array<String> {
            return proxyItem.split(delimiter.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        }
    }
}
